**Connect With Me in Linkedin** :- https://www.linkedin.com/in/dheerajkumar1997/

## Feature Selection: Duplicated Features

**Duplicate Features** are the ones when both the features in essenece are identical. Duplicate features might be already present in the dataset or might be introduced using one-hot encoding.

In [1]:
# Import Dependencies
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

In [2]:
# Load Dataset
df = pd.read_csv('./dataset/Santander-Customer-Satisfaction-data/train.csv', nrows=15000)
df.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3   ...    \
0                        0                        0   ...     
1                        0                        0   ...     
2                        0                        0   ...     
3                        0                        0   ...     
4                        0                        0   ...     

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   
2                      0.0                      0.0                     0.0   
3                      0.0                      0.0                     0.0   
4                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var44_ult1  saldo_medio_var44_ult3          var38  TARGET  
0                     0.0                     0.0   39205.170000       0  
1                     0.0                     0.0   49278.030000       0  
2                     0.0                     0.0   67333.770000       0  
3                     0.0                     0.0   64007.970000       0  
4                     0.0                     0.0  117310.979016       0  

[5 rows x 371 columns]

In [3]:
X = df.drop(['TARGET'], axis=1)
X.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3      ...        \
0                        0                        0      ...         
1                        0                        0      ...         
2                        0                        0      ...         
3                        0                        0      ...         
4                        0                        0      ...         

   saldo_medio_var29_ult3  saldo_medio_var33_hace2  saldo_medio_var33_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var33_ult1  saldo_medio_var33_ult3  saldo_medio_var44_hace2  \
0                     0.0                     0.0                      0.0   
1                     0.0                     0.0                      0.0   
2                     0.0                     0.0                      0.0   
3                     0.0                     0.0                      0.0   
4                     0.0                     0.0                      0.0   

   saldo_medio_var44_hace3  saldo_medio_var44_ult1  saldo_medio_var44_ult3  \
0                      0.0                     0.0                     0.0   
1                      0.0                     0.0                     0.0   
2                      0.0                     0.0                     0.0   
3                      0.0                     0.0                     0.0   
4                      0.0                     0.0                     0.0   

           var38  
0   39205.170000  
1   49278.030000  
2   67333.770000  
3   64007.970000  
4  117310.979016  

[5 rows x 370 columns]

In [4]:
y = df['TARGET']
y.head()

0    0
1    0
2    0
3    0
4    0
Name: TARGET, dtype: int64

In [5]:
# Train Test Split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=101)
X_train.shape, y_train.shape, X_test.shape, y_test.shape

((10500, 370), (10500,), (4500, 370), (4500,))

In [6]:
data_t = X_train.T
data_t.head()

7495    1247     10126    10708    3527      12732  \
ID                       15020.0  2468.0  20317.0  21463.0  7058.00  25525.00   
var3                         2.0     2.0      2.0      2.0     2.00      2.00   
var15                       23.0    41.0     34.0     23.0    34.00     30.00   
imp_ent_var16_ult1           0.0     0.0      0.0      0.0     0.00      0.00   
imp_op_var39_comer_ult1      0.0     0.0      0.0      0.0   467.67    106.89   

                            9636     10027   1632     7177    ...      973    \
ID                       19352.00  20124.0  3237.0  14405.0   ...     1929.0   
var3                         2.00      2.0     2.0      2.0   ...        2.0   
var15                       40.00     41.0    27.0     23.0   ...       23.0   
imp_ent_var16_ult1           0.00      0.0     0.0      0.0   ...        0.0   
imp_op_var39_comer_ult1    668.79     63.0     0.0      0.0   ...        0.0   

                          4079     12363    6141     599      5695     8006   \
ID                       8173.0  24789.0  12346.0  1191.00  11427.0  16142.0   
var3                        2.0      2.0      2.0     2.00      2.0      2.0   
var15                      40.0     26.0     32.0    23.00     24.0     23.0   
imp_ent_var16_ult1          0.0      0.0      0.0     0.00      0.0      0.0   
imp_op_var39_comer_ult1   240.0      0.0      0.0    24.54    365.7      0.0   

                          1361    1547     13151  
ID                       2677.0  3057.0  26357.0  
var3                        2.0     2.0      2.0  
var15                      57.0    23.0     65.0  
imp_ent_var16_ult1          0.0     0.0      0.0  
imp_op_var39_comer_ult1     0.0     0.0      0.0  

[5 rows x 10500 columns]

In [7]:
# Search for Duplicated Columns
data_t.duplicated().sum()

99

In [8]:
# Visualize Duplicated Rows
data_t[data_t.duplicated()]

7495   1247   10126  10708  3527   12732  \
ind_var2                         0.0    0.0    0.0    0.0    0.0    0.0   
ind_var13_medio_0                0.0    0.0    0.0    0.0    0.0    0.0   
ind_var13_medio                  0.0    0.0    0.0    0.0    0.0    0.0   
ind_var18_0                      0.0    0.0    0.0    0.0    0.0    0.0   
ind_var18                        0.0    0.0    0.0    0.0    0.0    0.0   
ind_var26                        0.0    0.0    0.0    0.0    0.0    1.0   
ind_var25                        0.0    0.0    0.0    0.0    0.0    1.0   
ind_var27_0                      0.0    0.0    0.0    0.0    0.0    0.0   
ind_var28_0                      0.0    0.0    0.0    0.0    0.0    0.0   
ind_var28                        0.0    0.0    0.0    0.0    0.0    0.0   
ind_var27                        0.0    0.0    0.0    0.0    0.0    0.0   
ind_var29_0                      0.0    0.0    0.0    0.0    0.0    0.0   
ind_var29                        0.0    0.0    0.0    0.0    0.0    0.0   
ind_var32                        0.0    0.0    0.0    0.0    0.0    0.0   
ind_var34_0                      0.0    0.0    0.0    0.0    0.0    0.0   
ind_var34                        0.0    0.0    0.0    0.0    0.0    0.0   
ind_var37                        0.0    0.0    0.0    0.0    0.0    0.0   
ind_var40_0                      0.0    0.0    0.0    0.0    0.0    0.0   
ind_var40                        0.0    0.0    0.0    0.0    0.0    0.0   
ind_var41                        0.0    0.0    0.0    0.0    0.0    0.0   
ind_var39                        0.0    0.0    0.0    0.0    0.0    0.0   
ind_var44                        0.0    0.0    0.0    0.0    0.0    0.0   
ind_var46_0                      0.0    0.0    0.0    0.0    0.0    0.0   
ind_var46                        0.0    0.0    0.0    0.0    0.0    0.0   
num_var13_medio_0                0.0    0.0    0.0    0.0    0.0    0.0   
num_var13_medio                  0.0    0.0    0.0    0.0    0.0    0.0   
num_var18_0                      0.0    0.0    0.0    0.0    0.0    0.0   
num_var18                        0.0    0.0    0.0    0.0    0.0    0.0   
num_var26                        0.0    0.0    0.0    0.0    0.0    3.0   
num_var25                        0.0    0.0    0.0    0.0    0.0    3.0   
...                              ...    ...    ...    ...    ...    ...   
imp_amort_var34_hace3            0.0    0.0    0.0    0.0    0.0    0.0   
imp_amort_var34_ult1             0.0    0.0    0.0    0.0    0.0    0.0   
imp_var7_emit_ult1               0.0    0.0    0.0    0.0    0.0    0.0   
imp_reemb_var13_hace3            0.0    0.0    0.0    0.0    0.0    0.0   
imp_reemb_var17_hace3            0.0    0.0    0.0    0.0    0.0    0.0   
imp_reemb_var33_hace3            0.0    0.0    0.0    0.0    0.0    0.0   
imp_reemb_var33_ult1             0.0    0.0    0.0    0.0    0.0    0.0   
imp_trasp_var17_out_hace3        0.0    0.0    0.0    0.0    0.0    0.0   
imp_trasp_var17_out_ult1         0.0    0.0    0.0    0.0    0.0    0.0   
imp_trasp_var33_in_hace3         0.0    0.0    0.0    0.0    0.0    0.0   
imp_trasp_var33_out_hace3        0.0    0.0    0.0    0.0    0.0    0.0   
ind_var7_emit_ult1               0.0    0.0    0.0    0.0    0.0    0.0   
num_var2_0_ult1                  0.0    0.0    0.0    0.0    0.0    0.0   
num_var2_ult1                    0.0    0.0    0.0    0.0    0.0    0.0   
num_var7_emit_ult1               0.0    0.0    0.0    0.0    0.0    0.0   
num_meses_var13_medio_ult3       0.0    0.0    0.0    0.0    0.0    0.0   
num_reemb_var13_hace3            0.0    0.0    0.0    0.0    0.0    0.0   
num_reemb_var17_hace3            0.0    0.0    0.0    0.0    0.0    0.0   
num_reemb_var33_hace3            0.0    0.0    0.0    0.0    0.0    0.0   
num_reemb_var33_ult1             0.0    0.0    0.0    0.0    0.0    0.0   
num_trasp_var17_out_hace3        0.0    0.0    0.0    0.0    0.0    0.0   
num_trasp_var17_out_ult1         0.0    0.0    0.0    0.

In [9]:
# Duplicated Features
duplicated_features = data_t[data_t.duplicated()].index.values

In [10]:
duplicated_features

array(['ind_var2', 'ind_var13_medio_0', 'ind_var13_medio', 'ind_var18_0',
       'ind_var18', 'ind_var26', 'ind_var25', 'ind_var27_0',
       'ind_var28_0', 'ind_var28', 'ind_var27', 'ind_var29_0',
       'ind_var29', 'ind_var32', 'ind_var34_0', 'ind_var34', 'ind_var37',
       'ind_var40_0', 'ind_var40', 'ind_var41', 'ind_var39', 'ind_var44',
       'ind_var46_0', 'ind_var46', 'num_var13_medio_0', 'num_var13_medio',
       'num_var18_0', 'num_var18', 'num_var26', 'num_var25',
       'num_var27_0', 'num_var28_0', 'num_var28', 'num_var27',
       'num_var29_0', 'num_var29', 'num_var32', 'num_var34_0',
       'num_var34', 'num_var37', 'num_var40_0', 'num_var40', 'num_var41',
       'num_var39', 'num_var46_0', 'num_var46', 'saldo_var13_medio',
       'saldo_var18', 'saldo_var28', 'saldo_var27', 'saldo_var29',
       'saldo_var34', 'saldo_var40', 'saldo_var41', 'saldo_var46',
       'delta_imp_amort_var18_1y3', 'delta_imp_amort_var34_1y3',
       'delta_imp_reemb_var33_1y3', 'delta_imp_tra

In [11]:
#  Drop Duplicated Features
unique_df = data_t.drop_duplicates(keep='first').T
unique_df.shape

(10500, 271)

This shows that, we started from a dataset containing the duplicated features and hence the total number of columns 371 which after the duplicate features removal has come down to 271.

In [12]:
# Columns Removed from original Dataframe
removed_features = [col for col in df.columns if col not in unique_df.columns]
removed_features

['ind_var2',
 'ind_var13_medio_0',
 'ind_var13_medio',
 'ind_var18_0',
 'ind_var18',
 'ind_var26',
 'ind_var25',
 'ind_var27_0',
 'ind_var28_0',
 'ind_var28',
 'ind_var27',
 'ind_var29_0',
 'ind_var29',
 'ind_var32',
 'ind_var34_0',
 'ind_var34',
 'ind_var37',
 'ind_var40_0',
 'ind_var40',
 'ind_var41',
 'ind_var39',
 'ind_var44',
 'ind_var46_0',
 'ind_var46',
 'num_var13_medio_0',
 'num_var13_medio',
 'num_var18_0',
 'num_var18',
 'num_var26',
 'num_var25',
 'num_var27_0',
 'num_var28_0',
 'num_var28',
 'num_var27',
 'num_var29_0',
 'num_var29',
 'num_var32',
 'num_var34_0',
 'num_var34',
 'num_var37',
 'num_var40_0',
 'num_var40',
 'num_var41',
 'num_var39',
 'num_var46_0',
 'num_var46',
 'saldo_var13_medio',
 'saldo_var18',
 'saldo_var28',
 'saldo_var27',
 'saldo_var29',
 'saldo_var34',
 'saldo_var40',
 'saldo_var41',
 'saldo_var46',
 'delta_imp_amort_var18_1y3',
 'delta_imp_amort_var34_1y3',
 'delta_imp_reemb_var33_1y3',
 'delta_imp_trasp_var17_out_1y3',
 'delta_num_aport_var17_1y3

### Remove Duplicate Features from Large Datasets

In [13]:
df = pd.read_csv('./dataset/Santander-Customer-Satisfaction-data/train.csv', nrows=50000)
df.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3   ...    \
0                      0.0                      0.0   ...     
1                      0.0                      0.0   ...     
2                      0.0                      0.0   ...     
3                      0.0                      0.0   ...     
4                      0.0                      0.0   ...     

   saldo_medio_var33_hace2  saldo_medio_var33_hace3  saldo_medio_var33_ult1  \
0                      0.0                      0.0                     0.0   
1                      0.0                      0.0                     0.0   
2                      0.0                      0.0                     0.0   
3                      0.0                      0.0                     0.0   
4                      0.0                      0.0                     0.0   

   saldo_medio_var33_ult3  saldo_medio_var44_hace2  saldo_medio_var44_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var44_ult1  saldo_medio_var44_ult3          var38  TARGET  
0                     0.0                     0.0   39205.170000       0  
1                     0.0                     0.0   49278.030000       0  
2                     0.0                     0.0   67333.770000       0  
3                     0.0                     0.0   64007.970000       0  
4                     0.0                     0.0  117310.979016       0  

[5 rows x 371 columns]

In [14]:
X = df.drop(['TARGET'], axis=1)
X.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3      ...        \
0                      0.0                      0.0      ...         
1                      0.0                      0.0      ...         
2                      0.0                      0.0      ...         
3                      0.0                      0.0      ...         
4                      0.0                      0.0      ...         

   saldo_medio_var29_ult3  saldo_medio_var33_hace2  saldo_medio_var33_hace3  \
0                     0.0                      0.0                      0.0   
1                     0.0                      0.0                      0.0   
2                     0.0                      0.0                      0.0   
3                     0.0                      0.0                      0.0   
4                     0.0                      0.0                      0.0   

   saldo_medio_var33_ult1  saldo_medio_var33_ult3  saldo_medio_var44_hace2  \
0                     0.0                     0.0                      0.0   
1                     0.0                     0.0                      0.0   
2                     0.0                     0.0                      0.0   
3                     0.0                     0.0                      0.0   
4                     0.0                     0.0                      0.0   

   saldo_medio_var44_hace3  saldo_medio_var44_ult1  saldo_medio_var44_ult3  \
0                      0.0                     0.0                     0.0   
1                      0.0                     0.0                     0.0   
2                      0.0                     0.0                     0.0   
3                      0.0                     0.0                     0.0   
4                      0.0                     0.0                     0.0   

           var38  
0   39205.170000  
1   49278.030000  
2   67333.770000  
3   64007.970000  
4  117310.979016  

[5 rows x 370 columns]

In [15]:
y = df['TARGET']
y.head()

0    0
1    0
2    0
3    0
4    0
Name: TARGET, dtype: int64

In [16]:
# Train Test Split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=101)
X_train.shape, y_train.shape, X_test.shape, y_test.shape

((35000, 370), (35000,), (15000, 370), (15000,))

In [17]:
# check for duplicated features in the training set
duplicated_feat = []
for i in range(0, len(X_train.columns)):
    if i % 10 == 0:  # this helps me understand how the loop is going
        print(i)
 
    col_1 = X_train.columns[i]
 
    for col_2 in X_train.columns[i + 1:]:
        if X_train[col_1].equals(X_train[col_2]):
            duplicated_feat.append(col_2)
            
len(duplicated_feat)

0
10
20
30
40
50
60
70
80
90
100
110
120
130
140
150
160
170
180
190
200
210
220
230
240
250
260
270
280
290
300
310
320
330
340
350
360


659

In [18]:
# Drop Duplicate Features from Training and Test Set
X_train.drop(labels=duplicated_feat, axis=1, inplace=True)
X_test.drop(labels=duplicated_feat, axis=1, inplace=True)
 
X_train.shape, X_test.shape

c:\users\hare krishna\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
c:\users\hare krishna\appdata\local\programs\python\python36\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


((35000, 306), (15000, 306))

**Connect With Me in Linkedin** :- https://www.linkedin.com/in/dheerajkumar1997/